icicle_v3/backend/cpu/src/field/cpu_ntt.cpp

```diff
@@ -1,33 +1,44 @@
-
+#include "ntt_template.h"
 #include "icicle/ntt.h"
-#include "icicle/errors.h"
-#include "icicle/runtime.h"
-#include "icicle/utils/log.h"
-#include "ntt.template"
-
-#include "icicle/fields/field_config.h"
-
 using namespace field_config;
 using namespace icicle;
 
-REGISTER_NTT_BACKEND("CPU", (cpu_ntt<scalar_t, scalar_t>));
-#ifdef EXT_FIELD
-REGISTER_NTT_EXT_FIELD_BACKEND("CPU", (cpu_ntt<scalar_t, extension_t>));
-#endif // EXT_FIELD
+eIcicleError
+cpu_ntt_init_domain(const Device& device, const scalar_t& primitive_root, const NTTInitDomainConfig& config)
+{
+  auto err = ntt_template::cpu_ntt_init_domain(device, primitive_root, config);
+  return err;
+}
 
-template <typename S = scalar_t>
-eIcicleError cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config)
+eIcicleError cpu_ntt_release_domain(const Device& device)
 {
-  ICICLE_LOG_ERROR << "cpu_ntt_init_domain() not implemented";
-  return eIcicleError::API_NOT_IMPLEMENTED;
+  auto err = ntt_template::cpu_ntt_release_domain(device);
+  return err;
 }
 
-REGISTER_NTT_INIT_DOMAIN_BACKEND("CPU", (cpu_ntt_init_domain<scalar_t>));
+template <typename S, typename E>
+eIcicleError cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+{
+  auto err = ntt_template::cpu_ntt<S, E>(device, input, size, dir, config, output);
+  return err;
+}
 
-eIcicleError cpu_ntt_release_domain(const Device& device)
+eIcicleError cpu_ntt_ref(
+  const Device& device, const scalar_t* input, int size, NTTDir dir, NTTConfig<scalar_t>& config, scalar_t* output)
 {
-  ICICLE_LOG_ERROR << "cpu_ntt_release_domain() not implemented";
-  return eIcicleError::API_NOT_IMPLEMENTED;
+  auto err = ntt_template::cpu_ntt_ref<scalar_t>(device, input, size, dir, config, output);
+  return err;
 }
 
-REGISTER_NTT_RELEASE_DOMAIN_BACKEND("CPU", cpu_ntt_release_domain);
\ No newline at end of file
+REGISTER_NTT_INIT_DOMAIN_BACKEND("CPU", (cpu_ntt_init_domain));
+REGISTER_NTT_INIT_DOMAIN_BACKEND("CPU_REF", (cpu_ntt_init_domain));
+
+REGISTER_NTT_RELEASE_DOMAIN_BACKEND("CPU", cpu_ntt_release_domain);
+REGISTER_NTT_RELEASE_DOMAIN_BACKEND("CPU_REF", cpu_ntt_release_domain);
+
+REGISTER_NTT_BACKEND("CPU", (cpu_ntt<scalar_t, scalar_t>));
+REGISTER_NTT_BACKEND("CPU_REF", (cpu_ntt_ref));
+
+#ifdef EXT_FIELD
+REGISTER_NTT_EXT_FIELD_BACKEND("CPU", (cpu_ntt<scalar_t, extension_t>));
```

Comment:
probably this case also needs to be registered for CPU_REF.

Reply:
Fixed